zulip/zulip
# input pill: Refactor to move custom attributes into relevant modules

This PR should have no visual changes to users, it's purely a refactor.

This change had two main motivators:

(1) We were putting a bunch of custom logic in `input_pill`, especially for display. This PR moves that to the relevant modules.

(2) `display_value` was being used for both the pill identifying "value" and for parsing the value a pill represented. This led to some hacks in the search pill code where we adjusted `display_value` just before displaying it. This PR removes `display_value` altogether and replaces it with relevant identifying information for each pill. Display values are calculated when we need them based on that identifying information. Maybe we want to tweak this to only generate `display_value` once, but I think the current state of the PR is fine since we don't display that pills in large quantities or in quick succession.

(3) (from a comment on the [search pills PR](https://github.com/zulip/zulip/pull/26803#issuecomment-2219032782)): The + format for strings with spaces in them should only be present in the code for generating a URL and parsing one; we should be storing the actual name in the pill object, not the URL-encoded name.

**Unit testing**

I removed some tests from `input_pill` that I haven't put elsewhere yet. If it seems like a good idea to add tests (especially `user_pill`) I can work on that.

**Manual testing**

search: DM (I also tested that the message view updates correctly)

<img width="227" alt="image" src="https://github.com/user-attachments/assets/8d4edf6f-8fce-450b-8411-08a244db21c9">

search: group DM (I also tested that the message view updates correctly)

<img width="461" alt="image" src="https://github.com/user-attachments/assets/b4c5eeb3-750d-4bb1-b67c-99703a9bebef">

search: DM with deactivated user (I also tested that the message view updates correctly)

<img width="323" alt="image" src="https://github.com/user-attachments/assets/bc1ee375-ce6c-42f4-8094-dfc697f26ddc">

search: DM with user with status_emoji_info (I also tested that the message view updates correctly)

<img width="276" alt="image" src="https://github.com/user-attachments/assets/959a5831-6807-499a-9a79-7c8f3bbc0634">

compose: deactivated user

<img width="360" alt="image" src="https://github.com/user-attachments/assets/5b5358ec-f12d-4af8-9377-7ed08c4765d4">

compose: user with guest indicator

<img width="381" alt="image" src="https://github.com/user-attachments/assets/067109a6-9573-4f14-9aa3-d048f2d13a84">

compose: user with status_emoji_info (I also tested that the message sends)

<img width="372" alt="image" src="https://github.com/user-attachments/assets/c6d74f74-d97b-4f46-8b28-638e34e22fb4">

stream add subscribers: user, usergroup, and channel (I also tested that adding works)

<img width="359" alt="image" src="https://github.com/user-attachments/assets/99d9b2e7-9ebc-4e50-b9cf-b91f480ce8a9">

stream add subscribers: user with status_emoji_info (it's expected that the emoji doesn't show)

<img width="209" alt="image" src="https://github.com/user-attachments/assets/247b0c87-9eaf-49aa-a3fc-e1bbc0539290">

invite: emails (I didn't test that the email was sent, only the visual state)

<img width="503" alt="image" src="https://github.com/user-attachments/assets/5ee1260f-82b8-4af2-ac68-6e286be4c7b1">

invite: stream picker  (I didn't test that the user was subscribed, only the visual state)

<img width="499" alt="image" src="https://github.com/user-attachments/assets/be27eff1-8bc0-4f03-9d47-f0ddfa6be546">
